drivers: espi_saf: Add Microchip MEC172x eSPI SAF version 2 driver

jvasanth1 opened the pull request:
Microchip MEC172x has a modified eSPI SAF hardware implementation. Hardware changes include multiple clock dividers for each SPI flash device and data transfer using QMSPI local DMA.

Other fixes as well:
1. boards: mec172xevb: Fix cmake no sources warning 
2. MEC172x adjust clock based on OTP 
3. MEC172x QMSPI clock fix
4. Set default interrupt priority to 3 for all Microchip MEC172x eSPI host child devices 
 (except the UART's which are set to 1)

jvasanth1 committed:
boards: mec172xevb: Fix cmake no sources warning

Microchip MEC172x EVB no longer has any C files in board
folder. Remove the cmake library rule as it it causing
a build warning due to no source files.

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 committed:
drivers: clock-control: Microchip MEC172x adjust clock based on OTP

Microchip MEC172x CPU and fast peripheral (QMSPI and PK) are
clock source is based upon an OTP setting. Add logic to adjust
clock source based on OTP value. If the OTP value is ever changed
this fix will allow calcluation of correct clock rate.

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 committed:
drivers: SPI: MEC172x QMSPI clock fix

Microchip MEC172x QMSPI expanded its clock divider register
field from 8 to 16 bits. QMSPI source clock is on the fast
peripheral domain therefore get the frequency from the clock
control driver.

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 committed:
drivers: espi: Microchip MEC172x child host device interrupt priorities

Set default interrupt priority to 3 for all Microchip MEC172x eSPI
host child devices except the UART's which are set to 1.
The espi peripherals don't require the maximum priority hence they
are being made uniform and a lower priority 3.

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 committed:
soc: microchip_mec: Prepare for MEC172x SAF version 2

Microchip MEC172x eSPI SAF has significant hardware changes
requiring a new SAF configuration structure. In preparation
for the MEC172x we move the current SAF header out of common
to the MEC1501 subfolder, remove an unused and empty common
header and remove includes of the moved headers from MEC172x.

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 committed:
drivers: espi_saf: Add Microchip MEC172x eSPI SAF version 2 driver

Microchip MEC172x has a modified eSPI SAF hardware implementation.
Hardware changes include multiple clock dividers for each SPI
flash device and data transfer using QMSPI local DMA.
espi reset interrupt is made a higer priority in MEC172x devicetree
because espi reset event resets all espi hardware and we don't
to want to service any other espi interrupt blocks when espi reset
occurs.

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

albertofloyd commented on drivers/espi/CMakeLists.txt:
```diff
@@ -7,7 +7,8 @@ zephyr_library_sources_ifdef(CONFIG_ESPI_NPCX		espi_npcx.c)
-zephyr_library_sources_ifdef(CONFIG_ESPI_SAF		espi_saf_mchp_xec.c)
-zephyr_library_sources_ifdef(CONFIG_ESPI_XEC_V2		espi_mchp_xec_v2.c)
+zephyr_library_sources_ifdef(CONFIG_ESPI_SAF_XEC	espi_saf_mchp_xec.c)
+zephyr_library_sources_ifdef(CONFIG_ESPI_XEC_V2         espi_mchp_xec_v2.c)
```
nit. spaces instead of tabs like rest of drivers

jvasanth1 replied:
updated

albertofloyd commented on drivers/espi/espi_saf_mchp_xec_v2.c:
```diff
@@ -0,0 +1,1169 @@
+	return 0;
+}
+
+/* !!!! TODO !!!!
```
Remove comment since the espi header update is part of this PR.
Add the eSPI header update in his own commit, prior to this one

jvasanth1 replied:
removed comment

albertofloyd commented on drivers/espi/Kconfig:
```diff
@@ -178,4 +176,20 @@ config ESPI_OOB_CHANNEL_RX_ASYNC
+config ESPI_SAF
+	bool "ESPI SAF driver"
+	depends on ESPI_FLASH_CHANNEL
+	default n
```
default n is implicit, so not required

albertofloyd commented on drivers/espi/Kconfig.xec:
```diff
@@ -54,18 +60,16 @@ config ESPI_FLASH_BUFFER_SIZE
+	depends on SOC_SERIES_MEC172X
 	help
-	  Driver initialization priority for eSPI SAF driver.
+	  Enable the Microchip XEC ESPI driver for MEC172x series.
```
Same here indicate SAF eSPI driver

albertofloyd commented on drivers/espi/espi_saf_mchp_xec_v2.c:
```diff
@@ -0,0 +1,1169 @@
+	regs->SAF_CS_OP[cs].OPA = fcfg->opa;
+	regs->SAF_CS_OP[cs].OPB = fcfg->opb;
+	regs->SAF_CS_OP[cs].OPC = fcfg->opc;
+	regs->SAF_CS_OP[cs].OP_DESCR = (uint32_t)fcfg->cs_cfg_descr_ids;
```
Notice OPD registers are not yet updated, will be this added later on?

jvasanth1 replied:
will add update to OPD register as well.

jvasanth1 replied:
yes, not updating OPD currently.

jvasanth1 commented on drivers/espi/Kconfig.xec:
```diff
@@ -54,18 +60,16 @@ config ESPI_FLASH_BUFFER_SIZE
 	help
-	  Enable Slave Attached Flash eSPI driver. SAF depends upon ESPI XEC driver
-	  and flash channel.
+	  Enable the Microchip XEC ESPI driver for MEC15xx family.
```
updated

albertofloyd commented on drivers/espi/espi_saf_mchp_xec_v2.c:
```diff
@@ -0,0 +1,1164 @@
+	.manage_callback = espi_saf_xec_manage_callback,
+};
+
+static struct espi_saf_xec_data espi_saf_xec_data;
```
Use a different name for variable other than struct name. This shall fix the problem with checks failing

albertofloyd commented on drivers/espi/espi_saf_mchp_xec_v2.c:
```diff
@@ -0,0 +1,1164 @@
+{
+	clock_control_subsys_t clkss =
+		(clock_control_subsys_t)(MCHP_XEC_PCR_CLK_PERIPH_FAST);
+	uint32_t clock = 0u;
```
Rename variable "clock" to fix check violation

jvasanth1 replied:
ok

albertofloyd commented on drivers/spi/spi_xec_qmspi_ldma.c:
```diff
@@ -159,36 +162,42 @@ static void qmspi_reset(struct qmspi_regs *regs)
-	uint32_t fdiv, qmode;
+	clock_control_subsys_t clkss =
+		(clock_control_subsys_t)(MCHP_XEC_PCR_CLK_PERIPH_FAST);
+	uint32_t clock = MCHP_QMSPI_INPUT_CLOCK_FREQ_HZ;
```
Rename variable "clock" to fix check violation

jvasanth1 replied:
ok

jvasanth1 commented:
Hi @albertofloyd 

There are few Koncifgs missing the in the combined Kconfig.xec (specific to MEC17xx) and hence the below compliance check error. Shall I add them under ESPI_XEC_V2 ?

```
CONFIG_ESPI_PERIPHERAL_ACPI_EC_IBF_EVT_DATA drivers/espi/espi_mchp_xec_host_v2.c:431, drivers/espi/espi_mchp_xec_host_v2.c:450, drivers/espi/espi_mchp_xec_host_v2.c:586, drivers/espi/espi_mchp_xec_host_v2.c:598
CONFIG_ESPI_PERIPHERAL_KBC_IBF_EVT_DATA    drivers/espi/espi_mchp_xec_host_v2.c:194
CONFIG_ESPI_PERIPHERAL_KBC_OBE_CBK         drivers/espi/espi_mchp_xec_host_v2.c:236
CONFIG_ESPI_PERIPHERAL_XEC_MAILBOX         drivers/espi/espi_mchp_xec_host_v2.c:108, drivers/espi/espi_mchp_xec_host_v2.c:169
CONFIG_ESPI_XEC_PERIPHERAL_ACPI_SHD_MEM_SIZE drivers/espi/espi_mchp_xec_host_v2.c:101, drivers/espi/espi_mchp_xec_host_v2.c:692
CONFIG_ESPI_XEC_PERIPHERAL_HOST_CMD_PARAM_SIZE drivers/espi/espi_mchp_xec_host_v2.c:100, drivers/espi/espi_mchp_xec_host_v2.c:103, drivers/espi/espi_mchp_xec_host_v2.c:489, drivers/espi/espi_mchp_xec_host_v2.c:691, drivers/espi/espi_mchp_xec_host_v2.c:694, drivers/espi/espi_mchp_xec_host_v2.c:696, drivers/espi/espi_mchp_xec_host_v2.c:775
```

albertofloyd commented:
> ned Kconfig.xec (specific to MEC17xx) and hence the below compliance check error. Shall I add them under ESPI_XEC_V2 ?

Yes

albertofloyd commented on soc/arm/microchip_mec/mec172x/Kconfig.defconfig.mec172xnsz:
```diff
@@ -39,6 +39,10 @@ config ESPI_XEC_V2
 	default y
 	depends on ESPI
 
+config ESPI_SAF_XEC_V2
```
Similar  change is missing for mec152x
./soc/arm/microchip_mec/mec1501/Kconfig.defconfig.mec1501hsz

With updated KConfig.xec ESPI_SAF_XEC is never set when build MEC15xx boards

jvasanth1 replied:
fixed

mbolivar-nordic commented on drivers/spi/spi_xec_qmspi_ldma.c:
```diff
@@ -1105,6 +1118,13 @@ static int qmspi_xec_init(const struct device *dev)
 		return ret;
 	}
 
+	ret = clock_control_get_rate(DEVICE_DT_GET(MCHP_XEC_CLOCK_CONTROL_NODE),
```
Is there a missing error path here if device_is_ready() is false for this device?

jvasanth1 replied:
will add device_is_ready() check

mbolivar-nordic commented on dts/arm/microchip/mec172xnsz.dtsi:
```diff
@@ -874,12 +874,17 @@
 			reg-names = "io", "mem", "vw";
 			interrupts = <103 3>, <104 3>, <105 3>, <106 3>,
-				     <107 3>, <108 3>, <109 3>, <110 3>,
+				     <107 3>, <108 3>, <109 3>, <110 2>,
```
Why is this being changed here? It looks like it might be a separate bug fix. The commit log is not clear about why these devicetree changes are needed in this patch in general.

jvasanth1 replied:
we are making espi reset interrupt a higher priority. Will add this information in the commit.

mbolivar-nordic replied:
> Will add this information in the commit.

The updated commit log says:

> espi reset interrupt is made a higer priority in MEC172x devicetree.

This is an explanation of **what** changes are being made, which I can see from the diff.

This is not an explanation of **why** the changes are being made. Please note that I wrote:

> The commit log is not clear about why these devicetree changes are needed in this patch in general.

jvasanth1 replied:
add explanation of why in the commit log

mbolivar-nordic commented:
@jvasanth1 since this did not make it in time for RC1, you have two options:

1. wait until the v3.2 window opens up and get this merged
2. split the bug fixes into a separate PR to be merged for v3.1, with the new driver deferred to v3.2

jvasanth1 commented:
we can approve this PR now and wait for the next window for merging.

mbolivar-nordic commented on dts/bindings/espi/microchip,xec-espi-saf-v2.yaml:
```diff
@@ -0,0 +1,64 @@
+      description: mmio register space
+      required: true
+
+    girqs:
+      type: array
+      required: true
+      description: Array of encoded interrupt information
```
I'm not familiar with this area, but why is this in `girqs` instead of `interrupts`?

@albertofloyd is this a standard property for espi devicetree bindings? Can you point me to any references in upstream linux bindings or another similarly authoritative place to make it clearer what is going on?

jvasanth1 replied:
the girqs are aggregated interrupts from the mec which hook to individual NVIC. The driver for a peripheral needs to know what interrupt in the girq it is hooked to clear the interrupt status in the girq.

albertofloyd replied:
@mbolivar-nordic these are mchp interrupt specifics, not eSPI related

mbolivar-nordic commented on dts/bindings/espi/microchip,xec-espi-saf-v2.yaml:
```diff
@@ -0,0 +1,64 @@
+      required: true
+      description: Array of encoded interrupt information
+
+    pcrs:
+      type: array
+      required: true
+      description: Array of eSPI PCR register index and bit position
```
A similar question here; why is this using a nonstandard `pcrs` property instead of using `reg`?

jvasanth1 replied:
every peripheral block has a sleep enable and clock enable bits, these are the PCR (power clock reset block) regs. The driver needs to know their respective sleep enable and clock enable bits.

mbolivar-nordic commented on dts/bindings/espi/microchip,xec-espi-saf-v2.yaml:
```diff
@@ -0,0 +1,64 @@
+      description: hold off poll after suspend in 20 ns units
+      required: false
+
+    sus-rsm=interval:
```
This looks like a typo; surely you didn't mean to use an `=` sign here; did you?

jvasanth1 replied:
yes, it is a typo

jvasanth1 replied:
fixed

jvasanth1 committed:
drivers: spi: fix formatting in mec172x ldma driver

fix leading white space and use tabs instead of space
in mec172x spi ldma driver.

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 commented:
@mbolivar-nordic can you help to review the responses to your review comments?

MaureenHelm commented on dts/arm/microchip/mec172xnsz.dtsi:
```diff
@@ -825,7 +825,7 @@
 				compatible = "microchip,xec-espi-host-dev";
 				reg = <0x400f0000 0x200>;
-				interrupts = <60 0>;
+				interrupts = <60 3>;
```
The commit message doesn't explain why you're changing interrupt priorities.

jvasanth1 replied:
will add this in the commit message
"The espi peripherals don't require the maximum priority hence they are being made uniform and a lower priority."

jvasanth1 replied:
updated

MaureenHelm commented on drivers/espi/espi_saf_mchp_xec_v2.c:
```diff
@@ -7,15 +7,15 @@
+#include <zephyr/drivers/espi_saf.h>
+#include <zephyr/drivers/interrupt_controller/intc_mchp_xec_ecia.h>
+#include <zephyr/dt-bindings/interrupt-controller/mchp-xec-ecia.h>
+#include <zephyr/logging/log.h>
```
Please squash these changes into the commit that introduces this file

jvasanth1 replied:
squashed

jvasanth1 commented:
@MaureenHelm can you help to review and approve?

MaureenHelm commented on drivers/espi/espi_saf_mchp_xec_v2.c:
```diff
@@ -0,0 +1,1164 @@
+	.irq_info_list = espi_saf_xec_irq_info_0,
+};
+
+DEVICE_DT_INST_DEFINE(0, &espi_saf_xec_init, NULL,
+		      &espisaf_xec_data, &espisaf_xec_config, POST_KERNEL,
+		      CONFIG_ESPI_SAF_INIT_PRIORITY, &espi_saf_xec_driver_api);
```
Please use `DT_INST_FOREACH_STATUS_OKAY`

jvasanth1 replied:
It is a single espi saf endpoint. We don't envision multiple saf (slave attached flash) instance.

jvasanth1 replied:
updated

jvasanth1 commented on drivers/espi/Kconfig.xec:
```diff
@@ -55,6 +61,97 @@ config ESPI_FLASH_BUFFER_SIZE
+	depends on ESPI_PERIPHERAL_CHANNEL
+	help
+	  Enable EMI 2 interface accessible via Host I/O over the
+	  ESPI Peripheral Channel.
```
updated

jvasanth1 commented on drivers/espi/Kconfig.xec:
```diff
@@ -9,7 +9,13 @@ config ESPI_XEC
-if ESPI_XEC
+config ESPI_XEC_V2
+	bool "XEC Microchip ESPI V2 driver"
+	depends on SOC_SERIES_MEC172X
```
updated

jvasanth1 commented:
@MaureenHelm and @albertofloyd can you help to approve this PR?

jvasanth1 commented:
@MaureenHelm can you help to review and approve?
